### PR TITLE
Fix TestInspectExecId

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

With these changes, this test seems to work now 100% of the time on my Windows to Linux testing. This has been one of the flakiest tests in Windows CI to date.

### CODE REVIEW

**Comment by duglin** [comment]
Are you looking at the right test?  That timeout isn't for deleting stuff, its waiting for an exec to start. It really takes over minute?

**Comment by jhowardmsft** [comment]
@duglin Yup, really am looking at the right test. It depends on the state of the daemon previously. 
. First fails at 62s (I had the timeout at 60). After it's clean, it's sub 2 seconds

**Comment by jhowardmsft** [comment]
And here's an example of it failing in CI on third attempt. (But CI showing green --- different story for Jess). https://jenkins.dockerproject.org/job/Windows-PRs/8135/console

**Comment by jhowardmsft** [comment]
Here's the daemon debug log once it doesn't have to delete things (and takes sub 2s)

**Comment by duglin** [comment]
@jhowardmsft PR #14706 should fix it so that we see the timeout during the loop instead of the inspect() after the loop throwing an error.

However, that doesn't explain why you're seeing a timeout.

I wonder if it would help if we added a "docker events" to the test, temporary, and grab all of that output so that when we timeout we can see what was going on during that test?

**Comment by duglin** [comment]
#14722 should fix the real issue here. I don't think you need to increase the timeout any more but please rebase and try it out.

**Comment by jhowardmsft** [comment]
@duglin - I spent a few hours debugging this today and already figured this one :) However, there's at least one really subtle other bug in this test which only seems to affect Windows CLI, but I'm sure there's a second too. 'read a' sometimes (1 in 5 times) when running the test bombs out early as for some reason the daemon sees something on stdin, hence when the test sits in the loop, the exec has already exited, and the test fails. I'm occasionally seeing a failure later in the test too. Why the Windows CLI causes the daemon to see that, I'm not sure, yet just from the command prompt running docker exec...., it works 100% of the time. Will look more if I get a chance over the weekend.

**Comment by tiborvass** [comment]
@jhowardmsft any news on the root cause of why it can't handle 10seconds ?

**Comment by jhowardmsft** [comment]
@tiborvass - see my comment above re something being seen on stdin prematurely. The timeout is a red herring. There is another bug too. I haven't had time as yet to root cause though.

**Comment by cpuguy83** [comment]
Well, the test is ignoring the error from `cmd.StdinPipe()`.

Can we instead just have the test run `top`  instead of dealing with the stdin, which isn't what's under test here?

**Comment by duglin** [comment]
See https://github.com/docker/docker/pull/15266   based on my quick test I saw a hang w/o -t but w/ -t it seemed to work. So there may be a real issue here with tty's.

**Comment by LK4D4** [comment]
@duglin Yes, I think we saw something like this before. I thought maybe we missed `Flush` somewhere.

**Comment by jhowardmsft** [comment]
duglin: OK, I believe I have solved this now. No more flakiness :)  However, I think we probably still need an issue as to why exec.Command running exec -i containerid sh -c read a almost always exits immediately when run under docker test from a Windows machine, but does not when run from the command line. However, as this test isn't designed to actually be testing stdin handling, but instead verifying that ExecIDs are being set correctly, for now this is the right fix to this test.

**Comment by tiborvass** [comment]
@jhowardmsft @duglin thanks guys! can we also file an issue for the -i on windows?

**Comment by tiborvass** [comment]
PASS: docker_cli_exec_test.go:307: DockerSuite.TestInspectExecID	12.920s :)

**Comment by jhowardmsft** [comment]
I've had it running in a hard loop on my local setup for the past 20+ minutes. Every one has passed :)

**Comment by jhowardmsft** [comment]
Issue 15274 opened for the other problem.

**Comment by duglin** [comment]
holding my nose on the sleeps  :-)   but LGTM

File: integration-cli/docker_cli_exec_test.go [hunk 1]
```diff
@@ -308,7 +308,7 @@ func (s *DockerSuite) TestInspectExecID(c *check.C) {
 	out, _ := dockerCmd(c, "run", "-d", "busybox", "top")
 	id := strings.TrimSuffix(out, "\n")
 
-	out, err := inspectField(id, "ExecIDs")
+	out, err := inspectField(id, ".ExecIDs")
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: You don't need dot https://github.com/docker/docker/blob/master/integration-cli/docker_utils.go#L923

File: integration-cli/docker_cli_exec_test.go [hunk 2]
```diff
@@ -316,19 +316,22 @@ func (s *DockerSuite) TestInspectExecID(c *check.C) {
 		c.Fatalf("ExecIDs should be empty, got: %s", out)
 	}
 
-	// Start an exec, have it block waiting for input so we can do some checking
-	cmd := exec.Command(dockerBinary, "exec", "-i", id, "sh", "-c", "read a")
-	execStdin, _ := cmd.StdinPipe()
-
+	// Start an exec, have it block so we can do some checking
+	cmd := exec.Command(dockerBinary, "exec", "-i", id, "sleep", "10")
```
Comment (LK4D4) [inline integration-cli/docker_cli_exec_test.go]: why use sleep and not top?

File: integration-cli/docker_cli_exec_test.go (hunk 1 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_exec_test.go]: Hmmm. Not convinced there. I tried this without as it originally was and it fails. I needed to add the dot, same as @duglin did in his linked PR. I'm rebased off master from this afternoon.

File: integration-cli/docker_cli_exec_test.go (hunk 2 above)
Comment (jhowardmsft) [inline integration-cli/docker_cli_exec_test.go]: Anything that stays there and can be easily exited would work. Sending a control-c to quit top would be more code than using sleep that is guaranteed to exit.

**Comment by duglin** [comment]
Trying w/o the sleep/racey concerns: #15291

**Comment by jhowardmsft** [comment]
Closing in favour of #15291

**Comment by thaJeztah** [comment]
test was renamed to `TestExecInspectID`  in https://github.com/moby/moby/commit/a9379b4af227f1966b71ee255fc8bcf272ffe43c

**Pull request closed without merging**